smartcontract/service/wasmvm/runtime.go

```diff
@@ -18,115 +18,648 @@
-	}
+func (self *Runtime) TimeStamp(proc *exec.Process) uint64 {
+	self.checkGas(TIME_STAMP_GAS)
+	return uint64(self.Service.Time)
+}
 
-	item, err := vm.GetPointerMemory(params[0])
+func (self *Runtime) BlockHeight(proc *exec.Process) uint32 {
+	self.checkGas(BLOCK_HEGHT_GAS)
+	return self.Service.Height
+}
+
+func (self *Runtime) SelfAddress(proc *exec.Process, dst uint32) {
+	self.checkGas(SELF_ADDRESS_GAS)
+	selfaddr := self.Service.ContextRef.CurrentContext().ContractAddress
+	_, err := proc.WriteAt(selfaddr[:], int64(dst))
 	if err != nil {
-		return false, err
+		panic(err)
 	}
+}
 
-	context := this.ContextRef.CurrentContext()
-	txHash := this.Tx.Hash()
-	event.PushSmartCodeEvent(txHash, 0, event.EVENT_LOG, &event.LogEventArgs{TxHash: txHash, ContractAddress: context.ContractAddress, Message: string(item)})
-	vm.RestoreCtx()
+func (self *Runtime) CallerAddress(proc *exec.Process, dst uint32) {
+	self.checkGas(CALLER_ADDRESS_GAS)
+	if self.Service.ContextRef.CallingContext() != nil {
+		calleraddr := self.Service.ContextRef.CallingContext().ContractAddress
+		_, err := proc.WriteAt(calleraddr[:], int64(dst))
+		if err != nil {
+			panic(err)
+		}
+	} else {
+		_, err := proc.WriteAt(common.ADDRESS_EMPTY[:], int64(dst))
+		if err != nil {
+			panic(err)
+		}
+	}
 
-	return true, nil
 }
 
-func (this *WasmVmService) runtimeCheckSig(engine *exec.ExecutionEngine) (bool, error) {
-	vm := engine.GetVM()
-	envCall := vm.GetEnvCall()
-	params := envCall.GetParams()
-	if len(params) != 3 {
-		return false, errors.NewErr("[RuntimeCheckSig]parameter count error ")
+func (self *Runtime) EntryAddress(proc *exec.Process, dst uint32) {
+	self.checkGas(ENTRY_ADDRESS_GAS)
+	entryAddress := self.Service.ContextRef.EntryContext().ContractAddress
+	_, err := proc.WriteAt(entryAddress[:], int64(dst))
+	if err != nil {
+		panic(err)
 	}
-	pubKey, err := vm.GetPointerMemory(params[0])
+}
+
+func (self *Runtime) Checkwitness(proc *exec.Process, dst uint32) uint32 {
+	self.checkGas(CHECKWITNESS_GAS)
+	addrbytes := make([]byte, 20)
+	_, err := proc.ReadAt(addrbytes, int64(dst))
 	if err != nil {
-		return false, err
+		panic(err)
 	}
-	data, err := vm.GetPointerMemory(params[1])
+
+	address, err := common.AddressParseFromBytes(addrbytes)
 	if err != nil {
-		return false, err
+		panic(err)
 	}
-	sig, err := vm.GetPointerMemory(params[2])
+
+	if self.Service.ContextRef.CheckWitness(address) {
+		return 1
+	}
+	return 0
+}
+
+func (self *Runtime) Ret(proc *exec.Process, ptr uint32, len uint32) {
+	bs := make([]byte, len)
+	_, err := proc.ReadAt(bs, int64(ptr))
 	if err != nil {
-		return false, err
+		panic(err)
 	}
-	res := 0
-	err = signature.Verify(pubKey, data, sig)
-	if err == nil {
-		res = 1
+
+	self.Output = bs
+	proc.Terminate()
+}
+
+func (self *Runtime) Debug(proc *exec.Process, ptr uint32, len uint32) {
+	bs := make([]byte, len)
+	_, err := proc.ReadAt(bs, int64(ptr))
+	if err != nil {
+		panic(err)
```

Comment:
from the wasm integration proposal, debug feature is optional, so it can not introduce any observable side affects.